c7n/resources/s3.py

```diff
@@ -431,75 +490,172 @@ def get_permissions(cls):
+        self.augment_fields = set(self.detect_augment_fields())
+        # location is required for client construction
+        self.augment_fields.add('Location')
+        # custodian always returns tags
+        self.augment_fields.add('Tags')
+
+    def validate(self):
+        config = self.get_augment_config()
+        if isinstance(config, str) and config not in ('all', 'detect', 'none'):
+            raise PolicyValidationError(
+                "augment-keys supports 'all', 'detect', 'none' or list of keys found: %s" % config)
+        elif isinstance(config, list):
+            delta = set(config).difference([row[1] for row in S3_AUGMENT_TABLE])
+            if delta:
+                raise PolicyValidationError("augment-keys - found invalid keys: %s" % (list(delta)))
+        if not isinstance(config, (list, str)):
+            raise PolicyValidationError(
+                "augment-keys supports 'all', 'detect', 'none' or list of keys found: %s" % config)
+
+    def get_augment_config(self):
+        augment_config = None
+        for option in self.manager.data.get('query', []):
+            if option and option.get('augment-keys') is not None:
+                augment_config = option['augment-keys']
+        if augment_config is None:
+            augment_config = 'all'
+        return augment_config
+
+    def detect_augment_fields(self):
+        # try to detect augment fields required for the policy execution
+        # we want to avoid extraneous api calls unless they are being used by the policy.
+
+        detected_keys = []
+        augment_keys = [row[1] for row in S3_AUGMENT_TABLE]
+        augment_config = self.get_augment_config()
+
+        if augment_config == 'all':
+            return augment_keys
+        elif augment_config == 'none':
+            return []
+        elif isinstance(augment_config, list):
+            return augment_config
+
+        for f in self.manager.iter_filters():
+            fkey = None
+            if not isinstance(f, ValueFilter):
                 continue
-            else:
+
+            f = f.data
+            # type: value
+            if f.get('type', '') == 'value':
+                fkey = f.get('key')
+            # k: v dict
+            elif len(f) == 1:
+                fkey = list(f.keys())[0]
+            if fkey is None:  # pragma: no cover
+                continue
+
+            # remove any jmespath expressions
+            fkey = fkey.split('.', 1)[0]
+
+            # tags have explicit handling in value filters.
+            if fkey.startswith('tag:'):
+                fkey = 'Tags'
+
+            # denied methods checks get all keys
+            if fkey.startswith('c7n:DeniedMethods'):
+                return augment_keys
+
+            if fkey in augment_keys:
+                detected_keys.append(fkey)
+
+        return detected_keys
+
+    def get_client(self, region):
+        if region in self.region_clients:
```

Comment:
Can't we use the walrus operator now?

```suggestion
        if client := self.region_clients.get(region):
            return client
```